### ISSUE TITLE
[WalletManager] block too far in the future has arrived to the wallet

### ISSUE THREAD

**Issue by Aprogiena** [stratisproject/StratisBitcoinFullNode#488]
Mined on testnet with StratisD, when suddenly Consensus height stopped to increase and Headers went far in front of it. Wallet height was stopped 2 blocks behind:

```
Headers.Height:        122618   Headers.Hash:         d517459ca0a7d8eed6adff1e684aadf844e589cf51861b8b95c53e7b2b3f648d
Consensus.Height:      122586   Consensus.Hash:       4836bc89a3f6be0248f07d5531f811363eb0684748f4c8d3273ffc1f4624d647
BlockStore.Height:     122586   BlockStore.Hash:      4836bc89a3f6be0248f07d5531f811363eb0684748f4c8d3273ffc1f4624d647
Wallet.Height:         122584   Wallet.Hash:          d6efc1f9a8b0c06dbba117fe8025f3ca26a5197c0840cd2071f7862726305c17
```

And in logs I can see the reason:

```
[2017-09-23 11:19:29.4024 12] TRACE: Stratis.Bitcoin.Features.Wallet.WalletManager.ProcessBlock block notification - height: 122586, hash: 4836bc89a3f6be0248f07d5531f811363eb0684748f4c8d3273ffc1f4624d647, coin: Stratis
[2017-09-23 11:19:29.4189 12] DEBUG: Stratis.Bitcoin.Features.Consensus.ConsensusFeature.RunLoop Exception occurred in consensus loop: Stratis.Bitcoin.Features.Wallet.WalletException: block too far in the future has arrived to the wallet
   at Stratis.Bitcoin.Features.Wallet.WalletManager.ProcessBlock(Block block, ChainedBlock chainedBlock) in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.Wallet\WalletManager.cs:line 508
   at Stratis.Bitcoin.Features.Wallet.WalletSyncManager.ProcessBlock(Block block) in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.Wallet\WalletSyncManager.cs:line 143
   at Stratis.Bitcoin.Features.Wallet.Notifications.BlockObserver.OnNextCore(Block block) in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.Wallet\Notifications\BlockObserver.cs:line 24
   at System.Reactive.SafeObserver`1.OnNext(TSource value)
   at System.Reactive.Observer`1.OnNext(T value)
   at System.Reactive.SynchronizedObserver`1.OnNextCore(T value)
   at Stratis.Bitcoin.Signals.Signaler`1.Broadcast(T item) in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Signals\Signaler.cs:line 70
   at Stratis.Bitcoin.Signals.Signals.SignalBlock(Block block) in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Signals\Signals.cs:line 74
   at Stratis.Bitcoin.Features.Consensus.ConsensusFeature.RunLoop(CancellationToken cancellationToken) in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.Consensus\ConsensusFeature.cs:line 179
[2017-09-23 11:19:29.4189 12] FATAL: Stratis.Bitcoin.Features.Consensus.ConsensusFeature.RunLoop Consensus loop unhandled exception (Tip:122586)
[2017-09-23 11:19:29.4354 12] INFO: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass12_0.<StartAsync>b__0 Consensus Loop stopping
[2017-09-23 11:19:29.4354 12] FATAL: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass12_0.<StartAsync>b__0 Consensus Loop threw an unhandled exception
[2017-09-23 11:19:29.4354 12] DEBUG: Stratis.Bitcoin.Utilities.AsyncLoop+<>c__DisplayClass12_0.<StartAsync>b__0 Consensus Loop threw an unhandled exception: Stratis.Bitcoin.Features.Wallet.WalletException: block too far in the future has arrived to the wallet
   at Stratis.Bitcoin.Features.Wallet.WalletManager.ProcessBlock(Block block, ChainedBlock chainedBlock) in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.Wallet\WalletManager.cs:line 508
   at Stratis.Bitcoin.Features.Wallet.WalletSyncManager.ProcessBlock(Block block) in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.Wallet\WalletSyncManager.cs:line 143
   at Stratis.Bitcoin.Features.Wallet.Notifications.BlockObserver.OnNextCore(Block block) in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.Wallet\Notifications\BlockObserver.cs:line 24
   at System.Reactive.SafeObserver`1.OnNext(TSource value)
   at System.Reactive.Observer`1.OnNext(T value)
   at System.Reactive.SynchronizedObserver`1.OnNextCore(T value)
   at Stratis.Bitcoin.Signals.Signaler`1.Broadcast(T item) in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Signals\Signaler.cs:line 70
   at Stratis.Bitcoin.Signals.Signals.SignalBlock(Block block) in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Signals\Signals.cs:line 74
   at Stratis.Bitcoin.Features.Consensus.ConsensusFeature.RunLoop(CancellationToken cancellationToken) in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.Consensus\ConsensusFeature.cs:line 200
   at Stratis.Bitcoin.Features.Consensus.ConsensusFeature.<<Start>b__22_0>d.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.Consensus\ConsensusFeature.cs:line 114
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Stratis.Bitcoin.Utilities.AsyncLoop.<>c__DisplayClass12_0.<<StartAsync>b__0>d.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Utilities\AsyncLoop.cs:line 122
[2017-09-23 11:19:29.7310 9] TRACE: Stratis.Bitcoin.Features.BlockStore.BlockStoreSignaled.<RelayWorker>b__14_0 ()
[2017-09-23 11:19:29.7310 9] TRACE: Stratis.Bitcoin.Features.BlockStore.BlockRepository.ExistAsync (hash:'4836bc89a3f6be0248f07d5531f811363eb0684748f4c8d3273ffc1f4624d647')
[2017-09-23 11:19:29.7310 9] TRACE: Stratis.Bitcoin.Features.BlockStore.BlockRepository.ExistAsync (-)
```

Exception in consensus loop simply stopped consensus to go futher, so no new block were accepted.

**Comment by Aprogiena** [comment]
If the node is restarted, it sync no problem.

Strange thing is that the wallet was higher at some point:
```
[2017-09-23 11:19:28.3876 3] INFO: Stratis.Bitcoin.FullNode.<StartPeriodicLog>b__70_0 ======Node stats====== 09/23/2017 11:19:28 agent StratisBitcoin:1.0.2
Headers.Height:        122586   Headers.Hash:         4836bc89a3f6be0248f07d5531f811363eb0684748f4c8d3273ffc1f4624d647
Consensus.Height:      122586   Consensus.Hash:       4836bc89a3f6be0248f07d5531f811363eb0684748f4c8d3273ffc1f4624d647
BlockStore.Height:     122585   BlockStore.Hash:      99f16335629bc4f11ef485872840a79182def83de35dc131d6425979a687c551
Wallet.Height:         122585   Wallet.Hash:          99f16335629bc4f11ef485872840a79182def83de35dc131d6425979a687c551
```

**Comment by Aprogiena** [comment]
Could this be the reason?
```
[2017-09-23 11:19:18.3574 39] INFO: Stratis.Bitcoin.FullNode.<StartPeriodicLog>b__70_0 ======Node stats====== 09/23/2017 11:19:18 agent StratisBitcoin:1.0.2
Headers.Height:        122585   Headers.Hash:         d7074600f661e23a9eb75f44092fa2b0f07c54b49cdc8e17f4e9c348f8285360
Consensus.Height:      122585   Consensus.Hash:       d7074600f661e23a9eb75f44092fa2b0f07c54b49cdc8e17f4e9c348f8285360
BlockStore.Height:     122585   BlockStore.Hash:      99f16335629bc4f11ef485872840a79182def83de35dc131d6425979a687c551
Wallet.Height:         122585   Wallet.Hash:          d7074600f661e23a9eb75f44092fa2b0f07c54b49cdc8e17f4e9c348f8285360

```

**Comment by Aprogiena** [comment]
OK, here is more details about the issue. My node mined block 
```
Found new POS block hash 'd7074600f661e23a9eb75f44092fa2b0f07c54b49cdc8e17f4e9c348f8285360' at height 122585
```

And it seems that it was connected to two nodes 13.64.76.48 and 51.141.51.129. The first peer rejected that the second peer accepted it. Soon after that, the second peer stopped advancing in the chain (just as my node).

```
Node:::ffff:13.64.76.48:26178,           connected (outbound),      agent /Stratis:2.0.0.3/,  height=122641
Node:::ffff:51.141.51.129:26178,         connected (outbound),      agent StratisBitcoin:1.0.2,  height=122586
```

So this is weird, so I checked the logs even more to the history and found this:

```
[2017-09-23 11:18:03.1986 17]
Node:::ffff:13.64.76.48:26178,           connected (outbound),      agent /Stratis:2.0.0.3/,  height=122584
Node:::ffff:51.141.51.129:26178,         connected (outbound),      agent StratisBitcoin:1.0.2,  height=122528
```

This means the peer 51.141.51.129 was not advancing for some time.

**Comment by Aprogiena** [comment]
So I mined the block and probably the peer 13.64.76.48 also mined a block and because of that it rejected it. Then this peer mined another one and send it to me:

```
[2017-09-23 11:19:28.3876 3] INFO: Stratis.Bitcoin.FullNode.<StartPeriodicLog>b__70_0 ======Node stats====== 09/23/2017 11:19:28 agent StratisBitcoin:1.0.2
Headers.Height:        122586   Headers.Hash:         4836bc89a3f6be0248f07d5531f811363eb0684748f4c8d3273ffc1f4624d647
Consensus.Height:      122586   Consensus.Hash:       4836bc89a3f6be0248f07d5531f811363eb0684748f4c8d3273ffc1f4624d647
BlockStore.Height:     122585   BlockStore.Hash:      99f16335629bc4f11ef485872840a79182def83de35dc131d6425979a687c551
Wallet.Height:         122585   Wallet.Hash:          99f16335629bc4f11ef485872840a79182def83de35dc131d6425979a687c551
```

I can see my node checked that this new block is OK and reorged itself.

**Comment by Aprogiena** [comment]
More precisely:

1) We are at block 122584 and my network peer mined a new block 122585. 
2) BlockPuller was assigned the job to download it and it did.
```
[2017-09-23 11:18:09.6920 5] TRACE: Stratis.Bitcoin.Features.BlockStore.LoopSteps.BlockStoreInnerStepReadBlocks+<ExecuteAsync>d__2.MoveNext Puller provided block '99f16335629bc4f11ef485872840a79182def83de35dc131d6425979a687c551/122585', length 465.
[2017-09-23 11:18:09.6920 5] TRACE: Stratis.Bitcoin.Features.BlockStore.LoopSteps.BlockStoreInnerStepReadBlocks.AddDownloadedBlockToStore chainedBlockToStore='99f16335629bc4f11ef485872840a79182def83de35dc131d6425979a687c551/122585' added to the store.
[2017-09-23 11:18:09.6920 5] TRACE: Stratis.Bitcoin.Features.BlockStore.LoopSteps.BlockStoreInnerStepReadBlocks.ShouldBlocksBePushedToRepository Insert block size is 465 bytes, download stack contains 0 more blocks to download, last flush time was 115 ms ago.
[2017-09-23 11:18:09.6920 5] TRACE: Stratis.Bitcoin.Features.BlockStore.LoopSteps.BlockStoreInnerStepReadBlocks.ShouldBlocksBePushedToRepository pushBufferSizeReached=False / downloadStackEmpty=True / pushTimeReached=False
[2017-09-23 11:18:09.6920 5] TRACE: Stratis.Bitcoin.Features.BlockStore.BlockRepository.PutAsync (nextBlockHash:'99f16335629bc4f11ef485872840a79182def83de35dc131d6425979a687c551',blocks.Count:1)
[2017-09-23 11:18:09.6920 5] TRACE: Stratis.Bitcoin.Features.BlockStore.BlockRepository.PutAsync (-)
```
3) Wallet gets the block:
```
[2017-09-23 11:18:09.7106 12] TRACE: Stratis.Bitcoin.Features.BlockStore.BlockStoreSignaled.OnNextCore Block hash is '99f16335629bc4f11ef485872840a79182def83de35dc131d6425979a687c551'.

[2017-09-23 11:18:09.7106 12] TRACE: Stratis.Bitcoin.Features.Wallet.WalletManager.ProcessBlock block notification - height: 122585, hash: 99f16335629bc4f11ef485872840a79182def83de35dc131d6425979a687c551, coin: Stratis
[2017-09-23 11:18:09.7136 12] TRACE: Stratis.Bitcoin.Features.Wallet.WalletManager.ProcessTransaction transaction received - hash: 8afe7fa33b08ce9e07972cd96344f50d26ef298085ad1b591ed7b472b6269922, coin: Stratis
```
There is not enough logs in wallet, so we don't know what happened there, but just after this line in log from wallet we have mined a competitive block:
4)
```
[2017-09-23 11:18:09.7136 8] TRACE: Stratis.Bitcoin.Features.BlockStore.BlockRepository.PutAsync (nextBlockHash:'d7074600f661e23a9eb75f44092fa2b0f07c54b49cdc8e17f4e9c348f8285360',blocks.Count:1)
...
[2017-09-23 11:18:09.9342 8] INFO: Stratis.Bitcoin.Features.Miner.PosMinting.CheckStake ==================================================================
[2017-09-23 11:18:09.9342 8] INFO: Stratis.Bitcoin.Features.Miner.PosMinting.CheckStake Found new POS block hash 'd7074600f661e23a9eb75f44092fa2b0f07c54b49cdc8e17f4e9c348f8285360' at height 122585.
[2017-09-23 11:18:09.9342 8] INFO: Stratis.Bitcoin.Features.Miner.PosMinting.CheckStake ==================================================================

```

**Comment by Aprogiena** [comment]
So this means there is a problem (probably race condition) when we receive a new block from the network +- at the same time as we mine a new block. We need to inspect that case in all components that implement a code that reacts to such an event of new block being accepted.

**Comment by Aprogiena** [comment]
And we should also review PoW part as it is probably another place where a new block can be introduced to the system.

**Comment by Aprogiena** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [[WalletManager] block too far in the future has arrived to the wallet](https://trello.com/c/l4oN9CzA/251-walletmanager-block-too-far-in-the-future-has-arrived-to-the-wallet)

**Comment by dangershony** [comment]
Ho wowo yes this is a race condition, previously i solved it by updating the header and pushing the block to the downloaded queue in the puller, this meant that there was only one method of receiving a block  but I had issue with that approach (don't remember what)

Consensus should not stop syncing because of a bug in the wallet how ever until wallet is stand alone (like store) we want such bugs to float immediately. 

I thought the wallet can recover from such a state, I am not sure why you got this exception.
When you get this log
`  this.logger.LogWarning("Wallet is behind the best chain and the next block is not found in store");`
a loop of 10 attempts will try to fetch from store and if not continue execution, then try again in the next block received eventually the node should recover.

**Comment by Aprogiena** [comment]
I don't think we should focus on wallet recovery here. We should serialize the way of new block arrival to the system. That and the fact that the staking node should always prefer its own blocks over competitive remote blocks should mitigate any issues.

I'm not sure the puller is the place for such serialization, esp. if we wanted to prefer our own blocks. I know this extra requirement of preference makes it more complicated, but it is really how it should be done, otherwise the staking performance would not be optimal. What I mean by preference is that if it happened that you received a new block from network and just after that finished staking your own block, so that both new blocks are at the same height, you should use your own block and try to build on the top of it. I.e. you might be required to even replace the current tip that you processed already. Little awkward.

I wouldn't focus on wallet at all, I guess if the new block acceptance is serialized, everything else will be just fine. However, that does not mean that there are no other chain.Tip or consensusLoop.Tip races in the wallet itself. But that should be for another issue.

**Comment by dangershony** [comment]
What do you mean block serialize?
Hmm actually we should probably favor the received block over ours (with the same work), if we received it then most likely other peers have also already received it and our block will be rejected.

There can be a race condition when a block is pulled from the network and passed in consensus and in the same time the miner is pushing a block to the consensus code.
This can be mitigated maybe using a lock over the consensus validation code.

**Comment by Aprogiena** [comment]
What I mean by that is that we have multiple sources where the block can come to the system and we need a synchronized queue and a single consumer of the queue to accept that block. 

And I don't think such queue should be the puller's queue.

A lock is another possibility of how to achieve the safe behavior, but maybe it is less elegant solution.

**Comment by Aprogiena** [comment]
First idea that I have - new class that would contain - the new block that either arrived from puller or was mined or staked; then a flag where it came from (i.e. from remote or mined/staked locally) and an event. 

When you want to process the block, you create the new class and initialize it and then you put it to the queue. There is an exclusive consumer of this FIFO queue that takes the item and process it. For remotely arrived blocks, it would just do normal processing, but for newly mined blocks it would do little more processing (as is done in Pow/PosMining now). This processing is always done only if the block is a part of the longest chain. However, for locally mined blocks, it would also check if they should not replace the current tip - the situation is that we are mining block and just before we successfully mine it, a new block arrives from the network, so the newly minted block would not count (actually it would cause all sorts of problems in current implementation). With new implementation, it would be allowed to replace current tip (but not more than that). Once the block is processed, the event would be set.

The event is used as a signal for the miner that the processing is finished. Only when it is processed, the miner should continue.

Alternatively, we could use this exclusive task scheduler we have there to actually be used instead of the queue and the consumer - you would probably prefer that.

**Comment by dangershony** [comment]
I heard the `exclusive task scheduler` now you have my curiosity!

As for the replacing of a block (in case the network got a block before we did, that will be tricky as we will need to reorg that current block.

             E2
A B C D E1

E2 is our mined block, we would have to rewind E1 and then replace it with E2 and of course also update the chain, this is somewhat more complicated.

This can also be solved with a lock in the ConsensusLoop and a unified method in the ConsensusFeature. 

You are proposing to redesign the way we serve blocks to consensus?
This means that the LookaheadPuller will have its own loop to download blocks and push them to the consensus FIFO list.

We may not need a flag to specify origin just check that the new block is the next block (or current if we want to replace same height blocks) in ConcurrentChain.

It will be a bit of a challenge to use the `exclusive task scheduler` unless we don't await calls to the coinview db.

Let me think more in this design, my first intention was to put a lock in the ConsensusLoop class but a better design will be welcomed.

**Comment by Aprogiena** [comment]
I am proposing redesign because in current code there is repeated code basically on 3 places - as we have 3 possible sources of new blocks. Instead of the whole logic being copied to 3 places (and potentially executed in parallel), we should have a single line "heyNewBlock(block)" which should handle everything and correctly in a serial manner. This thing should care about the whole logic and about the rewinding if necessary. 

Lock is not good enough as it would just make those 3 places more complex with even more repeated code. 

I agree that replacing of the tip in case a new mined tip arrives is a big complication, but it makes sense. Fortunately, this part of the code needs not to be perfect in performance, but rather perfect in being resistant to race conditions.

**Comment by Aprogiena** [comment]
Sorry, I've just realized that I've misunderstood your comment and I've also mixed up several things. So little more explanation. There are two problems that I would like to be fixed here:

1. there is repeated code in those different places that can introduce a new block to the system
2. those 3 places act in parallel and there is no sync while the new block is being inserted

The first problem can be solved by that `heyNewBlock` method. And the second problem can be solved with many different ways - exclusive consumer with queue, exclusive scheduler, or just exclusive lock. 

So yes, lock is definitely a solution for the second part. Once we have `heyNewBlock` method, simply a single lock there will make sure the internals are executed exclusively and that's all that is needed.
